Build failure: rustc 1.73.0 (cross-compilation)

### Steps To Reproduce

Tested with ab3ea20adfe05a2328b4991612912ef7db67af83.

Steps to reproduce the behavior:
1. build pkgs.pkgsCross.aarch64-multiplatform.rustc (also happens with armv7l-hf-multiplatform)

### Build log

```
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/build/rustc-1.73.0-src/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names -Ccodegen-units=10 -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_vendor,"unikraft")' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_os,"teeos")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa","mips32r6","mips64r6","csky")' -Zmacro-backtrace -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Csplit-debuginfo=off -Ctarget-feature=-crt-static -Cprefer-dynamic -Zinline-mir '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/1.73.0/")' --target x86_64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: Option 'sysroot' given more than once

Build completed unsuccessfully in 0:00:11
make: *** [Makefile:12: all] Error 1
```

Full log: https://gist.github.com/2xsaiko/09032d1ca0b05ae38b88c1a15c81e9ff

### Additional context

Might be related to #267003 but seems to be a different error.

### Notify maintainers

@Havvy @figsoda @Mic92 @tjni @winterqt @zowoq

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.63, NixOS, 23.11 (Tapir), 23.11.20231121.cccc85f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

### Priorities

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Bumping this, I was able to reproduce the bug on `aarch64-multiplatform` and `riscv64` (I haven't tested other architectures) on ab3ea20adfe05a2328b4991612912ef7db67af83 and 033b01ae61e269c1ebe768974ae5acc7959f794d :

```bash
> nix build github:NixOS/nixpkgs/033b01ae61e269c1ebe768974ae5acc7959f794d#legacyPackages.x86_64-linux.pkgsCross.riscv64.rustc                                                                                               Documents/nixpkgs
error: builder for '/nix/store/iln6qs99xg6g0h4symm67p70gs6b23c2-rustc-riscv64-unknown-linux-gnu-1.74.0.drv' failed with exit code 2;
       last 10 log lines:
       > Building stage0 library artifacts (x86_64-unknown-linux-gnu)
       > error: failed to run `rustc` to learn about target-specific information
       >
       > Caused by:
       >   process didn't exit successfully: `/build/rustc-1.74.0-src/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names -Ccodegen-units=10 -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_vendor,"unikraft")' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_os,"teeos","hurd")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa","mips32r6","mips64r6","csky")' -Zmacro-backtrace -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Csplit-debuginfo=off -Ctarget-feature=-crt-static -Cprefer-dynamic -Zinline-mir '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/1.74.0/")' --target x86_64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
       >   --- stderr
       >   error: Option 'sysroot' given more than once
       >
       > Build completed unsuccessfully in 0:00:38
       > make: *** [Makefile:19: all] Error 1
       For full logs, run 'nix log /nix/store/iln6qs99xg6g0h4symm67p70gs6b23c2-rustc-riscv64-unknown-linux-gnu-1.74.0.drv'.
error: 1 dependencies of derivation '/nix/store/1jgx7wc0i9hbdpywgg2sgqhmwbip11ci-rustc-wrapper-1.74.0.drv' failed to build
```

@2xsaiko could you maybe give some details on how you managed to fix the issue on your end ? Thanks.

--- a/pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix
+++ b/pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix
@@ -1,8 +1,6 @@
 { lib
 , fetchFromGitHub
 , hostPlatform
-, cargo
-, rustc
 , lld
 }:
 
@@ -24,12 +22,7 @@ let
     };
   };
 
-  # inherit (cross) rustPlatform;
-  # ^ breaks because we are doing a no_std embedded build with a custom sysroot,
-  # but the fast_cross rustc wrapper already passes a sysroot argument
-  rustPlatform = cross.makeRustPlatform {
-    inherit rustc cargo;
-  };
+  inherit (cross) rustPlatform;
 
 in
 

--- a/pkgs/build-support/rust/rustc-wrapper/rustc-wrapper.sh
+++ b/pkgs/build-support/rust/rustc-wrapper/rustc-wrapper.sh
@@ -1,6 +1,19 @@
 #!@shell@
 
-extraBefore=(@sysroot@)
+defaultSysroot=(@sysroot@)
+
+for arg; do
+    case "$arg" in
+        --sysroot)
+            defaultSysroot=()
+            ;;
+        --)
+            break
+            ;;
+    esac
+done
+
+extraBefore=("${defaultSysroot[@]}")
 extraAfter=($NIX_RUSTFLAGS)
 
 # Optionally print debug info.

